packages/host/app/resources/auto-attached-card.ts

```diff
@@ -26,94 +23,60 @@ interface Args {
-    const { topMostStackItems, attachedCards } = named;
+    const { topMostStackItems, attachedCardIds, removedCardIds } = named;
     if (this.operatorModeStateService.state.submode === Submodes.Code) {
       return; // Don't auto-attach cards in code mode
     }
-    this.updateAutoAttachedCardsTask.perform(topMostStackItems, attachedCards);
-  }
 
-  private updateAutoAttachedCardsTask = restartableTask(
-    async (
-      topMostStackItems: StackItem[],
-      attachedCards: CardDef[] | undefined,
-    ) => {
-      await Promise.all(topMostStackItems.map((item) => item.ready()));
-      if (this.stackItemsChanged(topMostStackItems)) {
-        // we must be sure to clear the lastRemovedCards state so cards can be auto-attached again
-        // note: if two of the same cards are opened on separate stack, one will be auto-attached.
-        // If one is removed from one of the stacks, the card WILL be auto-attached.
-        this.lastRemovedCards.clear();
-      }
-      this.cards.clear();
-      topMostStackItems.forEach((item) => {
-        if (!this.hasRealmURL(item) || isIndexCard(item)) {
-          return;
+    if (this.stackItemsChanged(topMostStackItems)) {
+      this.cardIds.clear();
+      for (let item of topMostStackItems) {
+        if (!item.url) {
+          continue;
         }
-        if (
-          this.isAlreadyAttached(item.card, attachedCards) ||
-          this.wasPreviouslyRemoved(item.card)
-        ) {
-          return;
+        if (removedCardIds?.includes(item.url)) {
+          continue;
         }
-        this.cards.add(item.card);
-      });
-      this.lastStackedItems = topMostStackItems;
-    },
-  );
+        if (attachedCardIds?.includes(item.url)) {
+          continue;
+        }
+        this.cardIds.add(item.url);
+        this.lastStackedItems = topMostStackItems;
+      }
+    }
+  }
 
-  stackItemsChanged(topMostStackItems: StackItem[]) {
+  private stackItemsChanged(topMostStackItems: StackItem[]) {
     if (topMostStackItems.length !== this.lastStackedItems.length) {
       return true;
     }
     for (let i = 0; i < topMostStackItems.length; i++) {
-      if (topMostStackItems[i].card.id !== this.lastStackedItems[i].card.id) {
+      if (topMostStackItems[i].url !== this.lastStackedItems[i].url) {
         return true;
       }
     }
     return false;
   }
-
-  private hasRealmURL(stackItem: StackItem) {
-    let realmURL = stackItem.card[stackItem.api.realmURL];
-    return Boolean(realmURL);
-  }
-
-  private isAlreadyAttached(
-    lastTopMostCard: CardDef,
-    attachedCards: CardDef[] | undefined,
-  ) {
-    if (attachedCards === undefined) {
-      return false;
-    }
-    return attachedCards?.some((c) => c.id === lastTopMostCard.id);
-  }
-
-  private wasPreviouslyRemoved(lastTopMostCard: CardDef) {
-    return this.lastRemovedCards.has(lastTopMostCard.id);
-  }
-
-  @action
-  onCardRemoval(card: CardDef) {
-    this.lastRemovedCards.add(card.id);
-  }
```

Comment:
This is very wrong. we should never mutate the state of a resource outside of the thunk.